app/controllers/products_controller.rb

```diff
@@ -2,27 +2,28 @@ class ProductsController < ApplicationController
   include ProductsHelper
 
   before_action :set_product, only: %i[ show edit update destroy ]
+  before_action :set_categories, only: [:new, :create, :edit, :update]
```

Comment:
is this needed for create or update?

Reply:
Yes this is required
If create or update fails, they render new and edit templates and they wont have @categories references when this is done